### DISCUSSION TITLE
Integration wants to re-auth every 2 oder 3 days

### DISCUSSION THREAD

**Discussion by DieterBurgy** [astrandb/miele#166] [General]
All works well, except for the fact, that after a few days (2 or 3 days) I have to re-authenticate again!
I get the hint for that, because the communication to any Miele device will not work any more.

In the home assistant device list, I click "re-auth" (I have a German system, so I don't know the exact English wording), and proceed to the Miele Auth website.
After connection to my account (and the correct LAN IP address of my home assistant) all works well again.

What I'm a little concerned about is the fact, that I cannot find any file named ".miele-token-cache". Neither within "/config", nor "/config/.storage"?

Any clues?

Thanks in advance!!!
Regards!

**Comment by astrandb** [comment]
The reauthentication request is caused by repeated failures to get data from the api. Probably due to a shaky connection. Personally I see this request with intervals of 1-3 months. 
The integration saves the tokens with the tools built into Home Assistant (in the config registry) so there is no .miele-token-cache. 
Many times it is enough to reload the integration instead of starting the reauthenticaton flow. 
It has been difficult to pinpoint the cause of this problem but I will keep my eyes open.

**Reply by astrandb** [reply]
@DieterBurgy 
Normally Miele lets the access token be vaild for a month. That means that if your turn off your Home Assistant instance for more than a month you will have reauthenticate with Miele cloud to get the integration up and running again. I guess this is not the case for you so there is probably sub-optimal error recovery somewhere in the integration.

FYI: The tokens are saved in `.storage/core.config_entries`

It would be very helpful if you could look into the full log next time this happens. Look for tracebacks or messages like "Token not valid..." or "Authentication failure when fetching data". The logged context can hopefully give me a hint of what is going on.

**Comment by DieterBurgy** [comment]
Good morning Ake!

Thank you very much for your quick answer!
I'll try to deliver all the information. But I have to admint, that I'm newbie with Home Assistant ;-)

I took a look at the log and filtered for "Miele":
2022-12-04 09:16:23.828 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration miele which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-04 09:49:41.931 ERROR (MainThread) [custom_components.miele.fan] Turn_off: 400 - Bad Request
2022-12-04 12:44:03.002 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-04 12:44:03.003 ERROR (MainThread) [custom_components.miele] Error fetching miele data:
2022-12-04 14:09:17.001 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-04 14:09:17.002 ERROR (MainThread) [custom_components.miele] Error fetching miele data:
2022-12-04 16:14:49.751 ERROR (MainThread) [custom_components.miele.light] Turn_on: 500 - Internal Server Error
2022-12-04 16:14:55.517 ERROR (MainThread) [custom_components.miele.light] Turn_on: 500 - Internal Server Error
2022-12-04 16:15:00.552 ERROR (MainThread) [custom_components.miele.light] Turn_off: 400 - Bad Request
2022-12-04 16:37:56.903 ERROR (MainThread) [custom_components.miele.fan] Turn_off: 400 - Bad Request
2022-12-04 17:06:09.001 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-04 17:06:09.002 ERROR (MainThread) [custom_components.miele] Error fetching miele data:
2022-12-04 19:34:19.002 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-04 19:34:19.003 ERROR (MainThread) [custom_components.miele] Error fetching miele data:
2022-12-05 06:43:28.001 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-05 06:43:28.003 ERROR (MainThread) [custom_components.miele] Error fetching miele data:

You're right. It looks as if there is some kind of "repeating" error. Could that be the cause?

Also I took a look at the file you mentioned: Yes, there is an entry from Miele. Please see the attached file (screenshot of terminal, since I cannot copy-paste).

![Screenshot 2022-12-05 080117](https://user-images.githubusercontent.com/119053572/205571076-a5a3f3d2-0519-4550-9bd3-cfe5e2afa8e0.png)

Is there sth wrong with the expire date?

Thanks again for your help!

Kind regards!

**Comment by astrandb** [comment]
Thank you, this is helpful. I will dive into the internal handling of these errors. I will probably need more detailed logs, but I will come back to you on how to capture them.

The `expires_at` converts to Tuesday 3 January 2023 15:19:16.183 so it seems to be correct.

I'll be back...

**Comment by astrandb** [comment]
Just to confirm: Was the integration asking for reauthentication when you captured the log this morning?
If not: I would appreciate if you could capture the same log as soon as possible after you see the reatuthentication request. You can email the log - without filter - directly to me at ake@strandberg.eu.

**Reply by DieterBurgy** [reply]
No, it was NOT asking, when I took the log this morning.

I will capture it again, as soon as the hint for re-auth will appear!

Thank you!

**Comment by joel-bourquard** [comment]
Hi Ake,
I'm having the same issue:
```
2022-12-07 09:36:51.000 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration miele which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-07 10:06:01.006 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-07 10:06:01.014 ERROR (MainThread) [custom_components.miele] Error fetching miele data: 
2022-12-07 11:05:20.004 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-07 11:05:20.009 ERROR (MainThread) [custom_components.miele] Error fetching miele data: 
2022-12-07 12:53:38.004 WARNING (MainThread) [custom_components.miele] Timeout during coordinator fetch
2022-12-07 12:53:38.009 ERROR (MainThread) [custom_components.miele] Error fetching miele data: 
2022-12-07 13:45:46.501 ERROR (MainThread) [custom_components.miele] Authentication failed while fetching miele data: Authentication failure when fetching data
```
I think the latest log entry (2022-12-07 13:45:46.501) corresponds to the moment the plugin asked for re-authentication.

My token from ".storage/core.config_entries" has
     "expires_at": 1672739149.6961784
which is January 3rd, 2023.

If you need any other info please let me know!

**Comment by astrandb** [comment]
Thanks, now I know where to look.

**Comment by rudizl** [comment]
Just to point, that yesterday evening I received a message from Miele, that today between 7 and 10  in the morning my time there will be server maintenance, so maybe that`s the reason of those errors? Now is 17:36.

**Reply by joel-bourquard** [reply]
I also noticed significant episodes of Miele downtime in the last week or so, which manifested itself with the status going from "off" to "not_connected".
Personally I haven't noticed a direct connection between these downtimes and the "re-authentication" issue but...

But maybe the "re-authentication issue" in Miele plugin could be triggered by a different kind of downtime/glitches on Miele's side, that only affects their authentication service? Just a wild guess...

**Reply by joel-bourquard** [reply]
Here's additional information that I think will be useful:

This time, I did *not* re-authenticate Miele.  I just restarted Home Assistant and the Miele integration is still working (door opening detected on my washing machine).   So, at least in my case, these re-auth prompts appear to be false alarms, maybe an occasionnal technical auth issue that is interpreted incorrectly?

**Reply by astrandb** [reply]
You don't even need to restart HA. It is enough to reload the integration. It will start again but the red card asking for re-authentication will not go away until you restart HA.
When the integration receives a HTTP code 401 from the Miele cloud a re-authentication request is triggered. Timeout errors are just logged and retried 60 sec later. Perhaps we could try to wait for 3 or 5 consecutive  401:s before triggering the re-authentication.

**Comment by astrandb** [comment]
It is obvious that the Miele servers responds with 401 now and then for no obvious reson. 
I have now implemented a retry mechanism for authentication errors (response code 401). 
I have been logging the responses in two systems for 48 hours. This morning I got one 401 error on one of the systems but the new fix made a retry instead of immediately triggering a re-authentication request. It recovered on the first retry so a user would barely have noticed anything.
This fix will be released soon together with some other optimizations.
#170

**Reply by joel-bourquard** [reply]
Awesome, thanks a lot! I'll upgrade to the new version tomorrow and let you know in case it happens again 😉
All the best

**Comment by TimMayle** [comment]
Hello guys,
I'm still struggling to get the integration to reconfigure. Got a simple "Error" message. ("Token not valid, trigger renewal")
I'm on 0.1.1.
Thanks in advance.
Btw, seems Miele upgraded their API to 1.0.6: https://www.miele.com/developer/news.html
Tim

**Reply by astrandb** [reply]
This is a slightly different flavour of the re-authenication request. It can only happen directly after start of Home Assistant. I will look over the retry mechanism here also.

Re: API 1.0.6, some of the changes are not working, I have reported to Miele but no response so far.

**Reply by astrandb** [reply]
Improved handling in https://github.com/astrandb/miele/pull/184

**Comment by TimMayle** [comment]
Hello,
Installled 0.1.2 today and still can't configure the integration.
Logs shows
`aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://api.mcs3.miele.com/thirdparty/token')`

It feels like if my credentials were not anymore valid.
Any hint?

**Reply by astrandb** [reply]
Yes, it sounds like your credentials are invalid. You can try to login on Miele’s portal https://www.miele.com/developer/swagger-ui/ If you credentials are accepted there, they should work in the integration also.

**Reply by TimMayle** [reply]
Login on miele web site works fine.

The interaction with Miele seems fine, I get the HA screen to link with my HA. then I got the error.
The '.storage/core.config_entries' file don't contains any miele token. This is strange.

I also got this message when restarting HA:
![image](https://user-images.githubusercontent.com/66377845/207910448-b4d2b212-81ab-40a4-8c1f-ae57cc038ba7.png)
I then removed my YAML configuration. But got same behaviour.

**Reply by astrandb** [reply]
This is odd!
It works for hundreds of users so there must some dirt that has slipped into your setup.

The fact that the config entry is lacking miele tokens is explained by that  the login/authentication process fails somewhere.

Password with special characters are treated differently by Miele depending on in what context they are used. So avoid these characters in  your password with Miele.

Take a look in `.storage/application_credentials` and make sure that the credentials from Miele are present and correct.

I suggest that you make note of the credentials and remove them via Settings->Integrations - three-dot menu -> Authentication credentials.

Remove Miele Integration from Settings-> Integrations

Restart HA

Verify that the url:s in Settings->System->Network are correct.

Add the integration back and follow the instuctions. You should get a little form from Home Assistant / Nabu Casa that wants you to confirm a redirect address. Make sure that it is correct in your setup. Sometimes you can get this form by going back to HA when the login screen from Miele has popped up.

Now I cannot come up with more... 😩

**Reply by TimMayle** [reply]
Thanks a lot,
You pointed to the right direction with the `application_credentials`.
In my `configuration.yaml` I had moved the id and password into the `secrets.yaml` file like this:
```
miele:
  client_id: !secret miele_client_id
```
The migration script converted it into:
`"client_id": "secret_miele_client_id"` or something similar.

After removing the credentials and restarting HA I could reauthenticate properly.
Thanks a lot!

**Comment by larsmarius87** [comment]
Hi! I've made an error when trying to add this integration, and used my email as client-id and password as clien-secret. And now i cant find any place to change it. After removing and reinstalling rebootin ect. it still shows the loginpage, but the url shows that i still use my email as client id. i cant get to the config page prompted on first time. Any suggestion to change it? 
I have tried the "other" miele inegration, and that works, with the client-id and secret added to config.yaml.

I hope somone can direct me in the right direction :)

**Comment by astrandb** [comment]
Make sure that the Miele credentials are removed from <url_to_your_ha>/config/application_credentials. Then remove and re-add the integration.

**Reply by larsmarius87** [reply]
Thanks! 
It almost worked, but now it just jumps to mieles webpage already logged in, but gives "Account linking failed!". I am guessing there is something i need to change there?

**Reply by larsmarius87** [reply]
Edit: got it working! Thanks :)
